File: backend/app/schemas/business.py
from datetime import datetime, time
from uuid import UUID

from pydantic import BaseModel, ConfigDict, field_validator, model_validator


class BusinessListItem(BaseModel):
    id: UUID
    name: str
    slug: str
    address: str | None = None
    phone: str | None = None
    slot_duration_minutes: int
    is_active: bool

    model_config = ConfigDict(from_attributes=True)


class BusinessScheduleOut(BaseModel):
    day_of_week: int
    start_time: time | None = None
    end_time: time | None = None
    is_closed: bool

    model_config = ConfigDict(from_attributes=True)


class BusinessScheduleUpdate(BaseModel):
    day_of_week: int
    start_time: time | None = None
    end_time: time | None = None
    is_closed: bool

    model_config = ConfigDict(from_attributes=True)

    @field_validator("day_of_week")
    @classmethod
    def validate_day_of_week(cls, v: int) -> int:
        if not (0 <= v <= 6):
            raise ValueError("day_of_week must be between 0 and 6")
        return v

    @model_validator(mode="after")
    def validate_schedule_times(self) -> "BusinessScheduleUpdate":
        if not self.is_closed:
            if self.start_time is None or self.end_time is None:
                raise ValueError("start_time and end_time are required for open days")
            if self.end_time <= self.start_time:
                raise ValueError("end_time must be strictly after start_time")
        return self


class BusinessDetail(BaseModel):
    id: UUID
    name: str
    slug: str
    address: str | None = None
    phone: str | None = None
    description: str | None = None
    working_start_time: time
    working_end_time: time
    slot_duration_minutes: int
    is_active: bool
    created_at: datetime
    updated_at: datetime
    schedules: list[BusinessScheduleOut] | None = None

    model_config = ConfigDict(from_attributes=True)

    @field_validator("schedules", mode="after")
    @classmethod
    def sort_schedules(cls, v: list[BusinessScheduleOut] | None) -> list[BusinessScheduleOut] | None:
        if v is not None:
            return sorted(v, key=lambda x: x.day_of_week)
        return v


class BusinessUpdate(BaseModel):
    name: str | None = None
    description: str | None = None
    phone: str | None = None
    address: str | None = None
    working_start_time: time | None = None
    working_end_time: time | None = None
    slot_duration_minutes: int | None = None
    schedules: list[BusinessScheduleUpdate] | None = None

    model_config = ConfigDict(from_attributes=True)

    @field_validator("schedules")
    @classmethod
    def validate_schedules_list(cls, v: list[BusinessScheduleUpdate] | None) -> list[BusinessScheduleUpdate] | None:
        if v is not None:
            if len(v) != 7:
                raise ValueError("Schedules list must contain exactly 7 days")
            days = [item.day_of_week for item in v]
            if len(set(days)) != 7 or not all(day in days for day in range(7)):
                raise ValueError("Schedules must cover all days from 0 to 6 with no duplicates")
        return v

File: mobile/lib/features/business/domain/models/business.dart
class Business {
  final String id;
  final String name;
  final String slug;
  final String? address;
  final String? phone;
  final String? description;
  final int slotDurationMinutes;
  final bool isActive;
  final String? workingStartTime;
  final String? workingEndTime;
  final String? createdAt;
  final String? updatedAt;

  const Business({
    required this.id,
    required this.name,
    required this.slug,
    this.address,
    this.phone,
    this.description,
    required this.slotDurationMinutes,
    required this.isActive,
    this.workingStartTime,
    this.workingEndTime,
    this.createdAt,
    this.updatedAt,
  });

  factory Business.fromJson(Map<String, dynamic> json) {
    final id = json['id'] as String?;
    final name = json['name'] as String?;
    final slug = json['slug'] as String?;
    final slotDurationMinutes = json['slot_duration_minutes'] as int?;
    final isActive = json['is_active'] as bool?;

    if (id == null) {
      throw const FormatException('Missing required field: id');
    }
    if (name == null) {
      throw const FormatException('Missing required field: name');
    }
    if (slug == null) {
      throw const FormatException('Missing required field: slug');
    }
    if (slotDurationMinutes == null) {
      throw const FormatException(
        'Missing required field: slot_duration_minutes',
      );
    }
    if (isActive == null) {
      throw const FormatException('Missing required field: is_active');
    }

    return Business(
      id: id,
      name: name,
      slug: slug,
      address: json['address'] as String?,
      phone: json['phone'] as String?,
      description: json['description'] as String?,
      slotDurationMinutes: slotDurationMinutes,
      isActive: isActive,
      workingStartTime: json['working_start_time'] as String?,
      workingEndTime: json['working_end_time'] as String?,
      createdAt: json['created_at'] as String?,
      updatedAt: json['updated_at'] as String?,
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'name': name,
      'slug': slug,
      'address': address,
      'phone': phone,
      'description': description,
      'slot_duration_minutes': slotDurationMinutes,
      'is_active': isActive,
      'working_start_time': workingStartTime,
      'working_end_time': workingEndTime,
      'created_at': createdAt,
      'updated_at': updatedAt,
    };
  }

  @override
  bool operator ==(Object other) {
    if (identical(this, other)) return true;
    return other is Business &&
        other.id == id &&
        other.name == name &&
        other.slug == slug &&
        other.address == address &&
        other.phone == phone &&
        other.description == description &&
        other.slotDurationMinutes == slotDurationMinutes &&
        other.isActive == isActive &&
        other.workingStartTime == workingStartTime &&
        other.workingEndTime == workingEndTime &&
        other.createdAt == createdAt &&
        other.updatedAt == updatedAt;
  }

  @override
  int get hashCode {
    return Object.hash(
      id,
      name,
      slug,
      address,
      phone,
      description,
      slotDurationMinutes,
      isActive,
      workingStartTime,
      workingEndTime,
      createdAt,
      updatedAt,
    );
  }
}
